ably/realtime_presence.go

```diff
@@ -11,25 +11,50 @@ import (
 // It allows entering, leaving and updating presence state for the current
 // client or on behalf of other client.
 type RealtimePresence struct {
-	mu             sync.Mutex
-	sync           sync.WaitGroup
+	mtx            sync.Mutex
+	syncMtx        sync.Mutex // synchronizes Get with sync
 	data           string
 	serial         string
 	subs           *subscriptions
 	channel        *RealtimeChannel
+	event          chan State
 	members        map[string]*proto.PresenceMessage
 	stale          map[string]struct{}
 	state          proto.PresenceState
 	syncInProgress bool
+	observedAttach bool
 	synced         bool
 }
 
 func newRealtimePresence(channel *RealtimeChannel) *RealtimePresence {
-	return &RealtimePresence{
+	pres := &RealtimePresence{
 		subs:    newSubscriptions(subscriptionPresenceMessages),
 		channel: channel,
+		event:   make(chan State, 1),
 		members: make(map[string]*proto.PresenceMessage),
 	}
+	pres.syncMtx.Lock() // initially we lock waiting for channel's attached state
+	channel.On(pres.event, StateChanAttached)
+	go pres.eventloop()
+	return pres
+}
+
+func (pres *RealtimePresence) eventloop() {
+	for event := range pres.event {
+		switch event.State {
+		case StateChanAttached:
+			pres.mtx.Lock()
+			if !pres.observedAttach && !pres.syncInProgress {
+				pres.syncMtx.Unlock() // channel is attached and there's no sync in progress
+				pres.observedAttach = true
```

Comment:
The `observedAttach` is there in order to differentiate two attach events - 1) an initial attach (right after the channel gets created), for which we want to block waiting for sync and 2) regular attach event.

But now I see that Get does not return error, when it's called on Detached / Failed channel. I'll address that and see if I can get rid of `observedAttach`.